File: split_data.py
import numpy as np

#movieData = np.loadtxt('./testdata.dta', dtype = 'str', delimiter = ' ')
#index = np.loadtxt('./index.txt', dtype = 'str', delimiter = ' ')

movieData = []
index = []

#Do not read all this data into memory.
movidData_f = open("./all.dta", "r")
  
#this is less bad, since the file is smaller, but still
index_f = open("./all.idx", "r")
for line in index_f:
  index.append(line.strip("\n"))
  
base = open("split_base.txt","w+")
valid = open("split_valid.txt","w+")
hidden = open("split_hidden.txt","w+")
probe = open("split_probe.txt","w+")
qual = open("split_qual.txt","w+")

#from googling "how to read to files at the same time python", I found code 
#like this:

#with open(file1) as f1, open(fil2) as f2:
#  for x, y in zip(f1, f2):
#     do stuff(x,y)

#this seems like the best way to accomplish this without reading 2 gigabytes
#into memory, but this should make the memory issue much less bad

i = 0
for line in movidData_f:
  idx = index[i]
  if idx=="1" :
    base.write(line)
  elif idx=="2" :
    valid.write(line)
  elif idx=="3" :
    hidden.write(line)
  elif idx=="4" :
    probe.write(line)
  else:
    qual.write(line)
  i = i+1;

base.close() 
valid.close()
hidden.close()
probe.close()
qual.close()

File: split_data.cpp
#include<iostream>
#include<fstream>
using namespace std;

int main(int argc, char **argv)

{
  string directory="./";
  int i,id;
  if(argc==2)
    {
      directory+=argv[1];
      directory+="/";
    }
  ifstream movieData(directory+"all.dta");
  ifstream idData(directory+"all.idx");
  int len_MovieData=0, len_IdData=0;
  string buffer;

  //split the files, quit if an invalid id is found
  ofstream base(directory+"split_base.txt");
  ofstream valid(directory+"split_valid.txt");
  ofstream hidden(directory+"split_hidden.txt");
  ofstream probe(directory+"split_probe.txt");
  ofstream qual(directory+"split_qual.txt");
  while(getline(movieData,buffer) && idData >> id)
    {
      switch(id)
        {
        case 1:
          base << buffer;
          base << endl;
          break;
        case 2:
          valid << buffer;
          valid << endl;
          break;
        case 3:
          hidden << buffer;
          hidden << endl;
          break;
        case 4:
          probe << buffer;
          probe << endl;
          break;
        case 5:
          qual << buffer;
          qual << endl;
          break;
        default:
          cout << "invalid ID: " << id << " at index " << i << endl;
          movieData.close();
          idData.close();
          base.close();
          valid.close();
          hidden.close();
          probe.close();
          qual.close();
          return 1;
        }

    }
  movieData.close();
  idData.close();
  base.close();
  valid.close();
  hidden.close();
  probe.close();
  qual.close();
  return 0;
}
